File: transpiler/src/test/java/com/here/ivi/api/generator/baseapi/TopologicalSortTest.java
/*
 * Copyright (C) 2017 HERE Global B.V. and its affiliate(s). All rights reserved.
 *
 * This software, including documentation, is protected by copyright controlled by
 * HERE Global B.V. All rights are reserved. Copying, including reproducing, storing,
 * adapting or translating, any or all of this material requires the prior written
 * consent of HERE Global B.V. This material also contains confidential information,
 * which may not be disclosed to others without prior written consent of HERE Global B.V.
 *
 */

package com.here.ivi.api.generator.baseapi;

import static org.junit.Assert.assertEquals;

import com.here.ivi.api.model.cppmodel.*;
import java.util.Arrays;
import java.util.Collection;
import java.util.List;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.Parameterized;

@RunWith(Parameterized.class)
public class TopologicalSortTest {
  private static final String TYPE_A = "A";
  private static final String TYPE_B = "B";
  private static final String TYPE_C = "C";
  private static final String FIRST_STRUCT_NAME = "StructA";
  private static final String SECOND_STRUCT_NAME = "StructB";
  private static final String THIRD_STRUCT_NAME = "StructC";

  private static final String ENUM_NAME = "Kind";
  private static final CppEnum ENUM = CppEnum.create(ENUM_NAME);
  private static final String TYPE_DEF_NAME = "shortcut";

  private final List<CppElement> elements;
  private final List<Integer> expectedOrder;

  public TopologicalSortTest(List<CppElement> elements, List<Integer> expectedOrder) {
    this.elements = elements;
    this.expectedOrder = expectedOrder;
  }

  @Parameterized.Parameters
  @SuppressWarnings({"PMD.ExcessiveMethodLength"})
  public static Collection<Object[]> testData() {
    return Arrays.asList(
        new Object[][] {
          // sortIndependentStructsKeepsSameOrder
          {
            Arrays.asList(
                createCppStruct(FIRST_STRUCT_NAME, TYPE_A, TYPE_B),
                createCppStruct(SECOND_STRUCT_NAME, TYPE_B, TYPE_C)),
            Arrays.asList(0, 1)
          },
          // sortDependentStructs
          {
            Arrays.asList(
                createCppStruct(FIRST_STRUCT_NAME, TYPE_A, SECOND_STRUCT_NAME),
                createCppStruct(SECOND_STRUCT_NAME, TYPE_B, TYPE_C)),
            Arrays.asList(1, 0)
          },
          // sortSortedDependentStructsKeepsSameOrder
          {
            Arrays.asList(
                createCppStruct(FIRST_STRUCT_NAME, TYPE_A, TYPE_B),
                createCppStruct(SECOND_STRUCT_NAME, TYPE_B, FIRST_STRUCT_NAME)),
            Arrays.asList(0, 1)
          },
          // sortMultipleStructsWithDependencies
          {
            Arrays.asList(
                createCppStruct(FIRST_STRUCT_NAME, SECOND_STRUCT_NAME, THIRD_STRUCT_NAME),
                createCppStruct(SECOND_STRUCT_NAME, TYPE_B, THIRD_STRUCT_NAME),
                createCppStruct(THIRD_STRUCT_NAME, TYPE_A, TYPE_B)),
            Arrays.asList(2, 1, 0)
          },
          // sortEnumWithStructDependingOnIt
          {
            Arrays.asList(createCppStruct(FIRST_STRUCT_NAME, TYPE_A, ENUM_NAME), ENUM),
            Arrays.asList(1, 0)
          },
          // enumWithStructNotDependingOnItKeepsSameOrder
          {
            Arrays.asList(createCppStruct(FIRST_STRUCT_NAME, TYPE_A, TYPE_B), ENUM),
            Arrays.asList(0, 1)
          },
          // enumWithUsingDependingOnIt
          {Arrays.asList(createUsing(ENUM_NAME), ENUM), Arrays.asList(1, 0)},
          // enumWithUsingNotDependingOnItKeepsSameOrder
          {Arrays.asList(createUsing(TYPE_A), ENUM), Arrays.asList(0, 1)},
          // enumWithConstantDependingOnIt
          {Arrays.asList(createConstant(ENUM_NAME), ENUM), Arrays.asList(1, 0)},
          // enumWithConstantNotDependingOnItKeepsSameOrder
          {Arrays.asList(createConstant(TYPE_A), ENUM), Arrays.asList(0, 1)},
          // constantDependingOnStruct
          {
            Arrays.asList(
                createConstant(FIRST_STRUCT_NAME),
                createCppStruct(FIRST_STRUCT_NAME, TYPE_A, TYPE_B)),
            Arrays.asList(1, 0)
          },
          // constantNotDependingOnStructKeepsSameOrder
          {
            Arrays.asList(
                createConstant(TYPE_A), createCppStruct(FIRST_STRUCT_NAME, TYPE_A, TYPE_B)),
            Arrays.asList(0, 1)
          },
          // constantDependingOnDefinition
          {Arrays.asList(createConstant(TYPE_DEF_NAME), createUsing(TYPE_A)), Arrays.asList(1, 0)},
          // constantNotDependingOnDefinitionKeepsSameOrder
          {Arrays.asList(createConstant(TYPE_B), createUsing(TYPE_A)), Arrays.asList(0, 1)},
          // usingDependingOnUsing
          {
            Arrays.asList(createUsing("anotherShortcut", TYPE_DEF_NAME), createUsing(TYPE_A)),
            Arrays.asList(1, 0)
          },
          // usingNotDependingOnUsingKeepsSameOrder
          {
            Arrays.asList(createUsing(TYPE_A), createUsing("anotherShortcut", TYPE_B)),
            Arrays.asList(0, 1)
          },
          // usingDependingOnStruct
          {
            Arrays.asList(
                createUsing(FIRST_STRUCT_NAME), createCppStruct(FIRST_STRUCT_NAME, TYPE_A, TYPE_B)),
            Arrays.asList(1, 0)
          },
          // usingNotDependingOnStructKeepsSameOrder
          {
            Arrays.asList(createUsing(TYPE_C), createCppStruct(FIRST_STRUCT_NAME, TYPE_A, TYPE_B)),
            Arrays.asList(0, 1)
          },
          // structDependingOnDefinition
          {
            Arrays.asList(
                createCppStruct(FIRST_STRUCT_NAME, TYPE_DEF_NAME, TYPE_B), createUsing(TYPE_A)),
            Arrays.asList(1, 0)
          },
          // structNotDependingOnDefinitionKeepsSameOrder
          {
            Arrays.asList(createCppStruct(FIRST_STRUCT_NAME, TYPE_A, TYPE_B), createUsing(TYPE_C)),
            Arrays.asList(0, 1)
          },
        });
  }

  private static CppComplexTypeRef createComplex(String name) {
    return new CppComplexTypeRef.Builder(name).build();
  }

  private static CppStruct createCppStruct(String name, String firstType, String secondType) {
    CppStruct cppStruct = new CppStruct(name);

    cppStruct.fields.add(new CppField(createComplex(firstType), "x"));
    cppStruct.fields.add(new CppField(createComplex(secondType), "y"));

    return cppStruct;
  }

  private static CppUsing createUsing(String typeName) {
    return createUsing(TYPE_DEF_NAME, typeName);
  }

  private static CppUsing createUsing(String name, String typeName) {
    CppTypeDefRef cppDefinition = new CppTypeDefRef(name, createComplex(typeName));
    return new CppUsing(name, cppDefinition);
  }

  private static CppConstant createConstant(String typeName) {
    String name = "fixed";
    return new CppConstant(name, createComplex(typeName), new CppValue());
  }

  @Test
  public void checkOrder() {
    List<CppElement> sortedElements = new TopologicalSort(elements).sort();

    assertEquals(elements.size(), sortedElements.size());

    for (int i = 0; i < elements.size(); ++i) {
      int index = expectedOrder.get(i);
      assertEquals(elements.get(index), sortedElements.get(i));
    }
  }
}

File: gluecodium/src/test/java/com/here/gluecodium/generator/cpp/TopologicalSortTest.kt
/*
 * Copyright (C) 2016-2020 HERE Europe B.V.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 * License-Filename: LICENSE
 */

package com.here.gluecodium.generator.cpp

import com.here.gluecodium.generator.cpp.TopologicalSortTestHelper.createConstant
import com.here.gluecodium.generator.cpp.TopologicalSortTestHelper.createConstantWithAliasType
import com.here.gluecodium.generator.cpp.TopologicalSortTestHelper.createPath
import com.here.gluecodium.generator.cpp.TopologicalSortTestHelper.createStruct
import com.here.gluecodium.generator.cpp.TopologicalSortTestHelper.createTypeAlias
import com.here.gluecodium.generator.cpp.TopologicalSortTestHelper.createTypeRef
import com.here.gluecodium.model.lime.LimeDirectTypeRef
import com.here.gluecodium.model.lime.LimeEnumeration
import com.here.gluecodium.model.lime.LimeFunction
import com.here.gluecodium.model.lime.LimeLambda
import com.here.gluecodium.model.lime.LimeNamedElement
import com.here.gluecodium.model.lime.LimeParameter
import com.here.gluecodium.model.lime.LimeReturnType
import com.here.gluecodium.model.lime.LimeStruct
import org.junit.Assert.assertEquals
import org.junit.Test
import org.junit.runner.RunWith
import org.junit.runners.Parameterized

@RunWith(Parameterized::class)
class TopologicalSortTest(
    @Suppress("UNUSED_PARAMETER") testName: String,
    private val elements: List<LimeNamedElement>,
    private val expectedOrder: List<Int>
) {

    @Test
    fun checkOrder() {
        val sortedElements = TopologicalSort(elements).sort()

        assertEquals(elements.size, sortedElements.size)

        elements.indices.forEach {
            val index = expectedOrder[it]
            assertEquals(elements[index], sortedElements[it])
        }
    }

    companion object {
        private const val TYPE_A = "A"
        private const val TYPE_B = "B"
        private const val TYPE_C = "C"
        private const val FIRST_STRUCT_NAME = "StructA"
        private const val SECOND_STRUCT_NAME = "StructB"
        private const val THIRD_STRUCT_NAME = "StructC"

        private const val ENUM_NAME = "Kind"
        private const val TYPE_DEF_NAME = "shortcut"

        private val LIME_ENUM = LimeEnumeration(createPath(ENUM_NAME))

        private val LIME_ALIAS = createTypeAlias(
            TYPE_DEF_NAME, createTypeRef(
                TYPE_A
            ))

        @JvmStatic
        @Parameterized.Parameters(name = "{0}")
        fun testData(): Collection<Array<Any>> =
            listOf(
                arrayOf(
                    "sortIndependentStructsKeepsSameOrder", listOf(
                        createStruct(FIRST_STRUCT_NAME, TYPE_A, TYPE_B),
                        createStruct(SECOND_STRUCT_NAME, TYPE_B, TYPE_C)
                    ), listOf(0, 1)
                ), arrayOf(
                    "sortDependentStructs", listOf(
                        createStruct(FIRST_STRUCT_NAME, TYPE_A, SECOND_STRUCT_NAME),
                        createStruct(SECOND_STRUCT_NAME, TYPE_B, TYPE_C)
                    ), listOf(1, 0)
                ), arrayOf(
                    "sortSortedDependentStructsKeepsSameOrder", listOf(
                        createStruct(FIRST_STRUCT_NAME, TYPE_A, TYPE_B),
                        createStruct(SECOND_STRUCT_NAME, TYPE_B, FIRST_STRUCT_NAME)
                    ), listOf(0, 1)
                ), arrayOf(
                    "sortMultipleStructsWithDependencies", listOf(
                        createStruct(FIRST_STRUCT_NAME, SECOND_STRUCT_NAME, THIRD_STRUCT_NAME),
                        createStruct(SECOND_STRUCT_NAME, TYPE_B, THIRD_STRUCT_NAME),
                        createStruct(THIRD_STRUCT_NAME, TYPE_A, TYPE_B)
                    ), listOf(2, 1, 0)
                ), arrayOf(
                    "sortEnumWithStructDependingOnIt",
                    listOf(createStruct(FIRST_STRUCT_NAME, TYPE_A, ENUM_NAME), LIME_ENUM),
                    listOf(1, 0)
                ), arrayOf(
                    "enumWithStructNotDependingOnItKeepsSameOrder",
                    listOf(createStruct(FIRST_STRUCT_NAME, TYPE_A, TYPE_B), LIME_ENUM),
                    listOf(0, 1)
                ), arrayOf(
                    "enumWithUsingDependingOnIt",
                    listOf(createTypeAlias(TYPE_DEF_NAME, createTypeRef(ENUM_NAME)), LIME_ENUM),
                    listOf(1, 0)
                ), arrayOf(
                    "enumWithUsingNotDependingOnItKeepsSameOrder",
                    listOf(LIME_ALIAS, LIME_ENUM),
                    listOf(0, 1)
                ), arrayOf(
                    "enumWithConstantDependingOnIt",
                    listOf(createConstant(ENUM_NAME), LIME_ENUM),
                    listOf(1, 0)
                ), arrayOf(
                    "enumWithConstantNotDependingOnItKeepsSameOrder",
                    listOf(createConstant(TYPE_A), LIME_ENUM),
                    listOf(0, 1)
                ), arrayOf(
                    "constantDependingOnStruct", listOf(
                        createConstant(FIRST_STRUCT_NAME),
                        createStruct(FIRST_STRUCT_NAME, TYPE_A, TYPE_B)
                    ), listOf(1, 0)
                ), arrayOf(
                    "constantNotDependingOnStructKeepsSameOrder", listOf(
                        createConstant(TYPE_A),
                        createStruct(FIRST_STRUCT_NAME, TYPE_A, TYPE_B)
                    ), listOf(0, 1)
                ), arrayOf(
                    "constantDependingOnDefinition",
                    listOf(createConstantWithAliasType(), LIME_ALIAS),
                    listOf(1, 0)
                ), arrayOf(
                    "constantNotDependingOnDefinitionKeepsSameOrder",
                    listOf(createConstant(TYPE_B), LIME_ALIAS),
                    listOf(0, 1)
                ), arrayOf(
                    "usingDependingOnUsing", listOf(
                        createTypeAlias(
                            "anotherShortcut",
                            LimeDirectTypeRef(createTypeAlias(TYPE_DEF_NAME, createTypeRef(TYPE_A)))
                        ),
                        LIME_ALIAS
                    ), listOf(1, 0)
                ), arrayOf(
                    "usingNotDependingOnUsingKeepsSameOrder",
                    listOf(LIME_ALIAS, createTypeAlias("anotherShortcut", createTypeRef(TYPE_B))),
                    listOf(0, 1)
                ), arrayOf(
                    "usingDependingOnStruct", listOf(
                        createTypeAlias(TYPE_DEF_NAME, createTypeRef(FIRST_STRUCT_NAME)),
                        createStruct(FIRST_STRUCT_NAME, TYPE_A, TYPE_B)
                    ), listOf(1, 0)
                ), arrayOf(
                    "usingNotDependingOnStructKeepsSameOrder", listOf(
                        createTypeAlias(TYPE_DEF_NAME, createTypeRef(TYPE_C)),
                        createStruct(FIRST_STRUCT_NAME, TYPE_A, TYPE_B)
                    ), listOf(0, 1)
                ), arrayOf(
                    "usingFunctionTypeRefDependingOnStruct", listOf(
                        createTypeAlias(
                            TYPE_DEF_NAME,
                            LimeDirectTypeRef(LimeLambda(createPath(FIRST_STRUCT_NAME)))
                        ),
                        createStruct(FIRST_STRUCT_NAME, TYPE_A, TYPE_B)
                    ), listOf(1, 0)
                ), arrayOf(
                    "structDependingOnDefinition",
                    listOf(
                        createStruct(FIRST_STRUCT_NAME, TYPE_DEF_NAME, TYPE_B),
                        LIME_ALIAS
                    ),
                    listOf(1, 0)
                ), arrayOf(
                    "structNotDependingOnDefinitionKeepsSameOrder", listOf(
                        createStruct(FIRST_STRUCT_NAME, TYPE_A, TYPE_B),
                        createTypeAlias(TYPE_DEF_NAME, createTypeRef(TYPE_C))
                    ), listOf(0, 1)
                ), arrayOf(
                    "sortStructsDependentThroughMethodReturnType", listOf(
                        LimeStruct(
                            path = createPath(FIRST_STRUCT_NAME),
                            functions = listOf(
                                LimeFunction(
                                    createPath("x"),
                                    returnType = LimeReturnType(createTypeRef(SECOND_STRUCT_NAME))
                                )
                            )
                        ),
                        createStruct(SECOND_STRUCT_NAME, TYPE_B, TYPE_C)
                    ), listOf(1, 0)
                ), arrayOf(
                    "sortStructsDependentThroughMethodParameter", listOf(
                        LimeStruct(
                            path = createPath(FIRST_STRUCT_NAME),
                            functions = listOf(
                                LimeFunction(
                                    createPath("x"),
                                    parameters = listOf(
                                        LimeParameter(
                                            createPath("x"),
                                            typeRef = createTypeRef(SECOND_STRUCT_NAME)
                                        ))
                                )
                            )
                        ),
                        createStruct(SECOND_STRUCT_NAME, TYPE_B, TYPE_C)
                    ), listOf(1, 0)
                ), arrayOf(
                    "sortStructDependentOnSelfDoesNotCrash", listOf(
                        LimeStruct(
                            path = createPath(FIRST_STRUCT_NAME),
                            functions = listOf(
                                LimeFunction(
                                    createPath("x"),
                                    returnType = LimeReturnType(createTypeRef(FIRST_STRUCT_NAME))
                                )
                            )
                        )
                    ), listOf(0)
                )
            )
    }
}
